File: app/src/main/java/com/vluver/beta/activities/PostActivity.java
package com.vluver.beta.activities;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserInfo;
import com.vluver.beta.R;
import com.vluver.beta.adapter.MultiAddImages;
import com.vluver.beta.model.AddMoreImages;
import com.vluver.beta.servicebackground.UploadImagesPost;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import gun0912.tedimagepicker.builder.TedImagePicker;
import gun0912.tedimagepicker.builder.listener.OnErrorListener;
import gun0912.tedimagepicker.builder.listener.OnMultiSelectedListener;
import gun0912.tedimagepicker.builder.type.MediaType;

import static com.vluver.beta.Direccion.urlgeneral;


public class PostActivity extends AppCompatActivity {
    ProgressDialog progressDialog;
    ImageView new_post_pictures;
    ImageView regresar,enviar,seleccionar_imagen;
    private ImageView user_image;
    EditText descripcion_post;
    boolean imageIsSet = false;
    private String userId;
    RequestQueue mQueue;

    //firebase
    private FirebaseAuth mAuth;
    public FirebaseUser currentUser;


    int PICK_IMAGE_MULTIPLE = 1;
    public RecyclerView mRVFish;
    public MultiAddImages mAdapter;
    List<AddMoreImages> dataa;
    GridLayoutManager mGridLayoutManager;

    ArrayList<Uri> encodedImageList;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_post);
        mAuth = FirebaseAuth.getInstance();
        mQueue = Volley.newRequestQueue(PostActivity.this);
        progressDialog = new ProgressDialog(this);
        new_post_pictures = (ImageView) findViewById(R.id.new_post_picture);
        regresar=(ImageView)findViewById(R.id.back);
        enviar=(ImageView) findViewById(R.id.send);
        descripcion_post =(EditText) findViewById(R.id.descripcion);
        user_image = (ImageView) findViewById(R.id.img_avatar);
        seleccionar_imagen = (ImageView) findViewById(R.id.new_post_submit);
        regresar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });
        mGridLayoutManager = new GridLayoutManager(PostActivity.this, 3);
        dataa=new ArrayList<>();
        encodedImageList = new ArrayList<>();
        mRVFish = (RecyclerView) findViewById(R.id.multimages);
        mAdapter = new MultiAddImages(PostActivity.this, dataa);
        mRVFish.setAdapter(mAdapter);
        mRVFish.setLayoutManager(mGridLayoutManager);
        seleccionar_imagen.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                TedImagePicker.with(PostActivity.this)
                        .max(99,"Ups! max 99 img")
                        .mediaType(MediaType.IMAGE).errorListener(new OnErrorListener() {
                    @Override
                    public void onError(@NotNull String s) {
                        imageIsSet = false;
                        Toast.makeText(PostActivity.this, "Error: "+s, Toast.LENGTH_SHORT).show();
                    }
                }).startMultiImage((OnMultiSelectedListener) uriList -> {
                            imageIsSet = true;
                            hideKeyboard(PostActivity.this);
                            for(Uri photos : uriList){
                                AddMoreImages addMoreImages = new AddMoreImages();
                                addMoreImages.mArrayUri = photos;
                                dataa.add(addMoreImages);
                                encodedImageList.add(photos);
                            }
                            mAdapter.notifyDataSetChanged();
                            mRVFish.setAdapter(mAdapter);

                        });
               /* TedImagePicker.with(PostActivity.this).max(99,"Max 99")
                        .start(new OnMultiSelectedListener() {
                            @Override
                            public void onSelected(@NotNull List<? extends Uri> uriList) {
                                imageIsSet = true;
                                for(Uri photos : uriList){
                                    AddMoreImages addMoreImages = new AddMoreImages();
                                    addMoreImages.mArrayUri = photos;
                                    dataa.add(addMoreImages);
                                    encodedImageList.add(String.valueOf(uriList));
                                }
                                mAdapter.notifyDataSetChanged();
                                mRVFish.setAdapter(mAdapter);
                                // multi
                            }
                        });*/
            }
        });

        enviar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (imageIsSet){



                    Intent mServiceIntent = new Intent(PostActivity.this, UploadImagesPost.class);
                    mServiceIntent.putExtra("userId", userId);
                    mServiceIntent.putParcelableArrayListExtra("mPhotos", encodedImageList);
                    mServiceIntent.putExtra("descripcion",descripcion_post.getText().toString().trim());
                    startService(mServiceIntent);
                    finish();
                }else{
                    if (descripcion_post.getText().toString().isEmpty()){
                        descripcion_post.setError("Agrega algo a tus publicacion...");
                    }else{
                        insertPostToVluver(userId,"null",descripcion_post.getText().toString());
                    }
                }


            }
        });
        getAvatar();

    }


    private void getAvatar(){
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        if (user != null) {
            for (UserInfo profile : user.getProviderData()) {
                if (profile.getProviderId().equals("facebook.com")){
                    String image = "https://graph.facebook.com/" + user.getProviderData().get(1).getUid() + "/picture?type=small";
                   // GlideLoadImages.loadAvatar(PostActivity.this,image,user_image);
                }else if (profile.getProviderId().equals("google.com")){
                    Toast.makeText(PostActivity.this, "avatar Google", Toast.LENGTH_SHORT).show();
                }
            }

        }
    }

    private void insertPostToVluver(String creator, String codeuniqueimages, String description){
        String url = urlgeneral+"insert/insertPost.php";
        StringRequest strReq = new StringRequest(Request.Method.POST,
                url, new Response.Listener<String>(){
            @Override
            public void onResponse(String response) {
                try {
                    JSONObject jsonObject = new JSONObject(response);
                    boolean error = jsonObject.getBoolean("error");
                    if (!error){
                        // progressDialog.dismiss();
                        Toast.makeText(PostActivity.this, "Publicacion subida correctamente", Toast.LENGTH_LONG).show();
                        //show_notification("Vluver", "Publicacion lista", 2 );
                        finish();
                    }else{
                        //progressDialog.dismiss();
                        String errorMsg = jsonObject.getString("error_msg");
                        Toast.makeText(PostActivity.this, ""+errorMsg, Toast.LENGTH_SHORT).show();
                    }

                } catch (JSONException e) {
                    // progressDialog.dismiss();
                    Toast.makeText(PostActivity.this, "Algo anda mal"+e, Toast.LENGTH_SHORT).show();
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {
                //progressDialog.dismiss();
                Toast.makeText(getApplicationContext(), error.getMessage(), Toast.LENGTH_LONG).show();
            }
        }) {

            @Override
            protected Map<String, String> getParams() {
                Map<String, String> params = new HashMap<String, String>();
                params.put("creator", creator);
                params.put("namecode",codeuniqueimages);
                params.put("description",description);
                return params;
            }


        };
        strReq.setRetryPolicy(new DefaultRetryPolicy(DefaultRetryPolicy.DEFAULT_TIMEOUT_MS, 5, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        mQueue.add(strReq);
    }


    public static void hideKeyboard(Activity activity) {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
        //Find the currently focused view, so we can grab the correct window token from it.
        View view = activity.getCurrentFocus();
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = new View(activity);
        }
        assert imm != null;
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }


    @Override
    protected void onStart() {
        super.onStart();
        currentUser = mAuth.getCurrentUser();
        userId = currentUser.getUid();
    }

    @Override
    protected void onDestroy() {
        mQueue.stop();
        mQueue = null;
        super.onDestroy();
    }
}

File: app/src/main/java/com/vluver/beta/activities/PostActivity.kt
package com.vluver.beta.activities

import android.Manifest
import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import com.android.volley.DefaultRetryPolicy
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.tangxiaolv.telegramgallery.GalleryActivity
import com.tangxiaolv.telegramgallery.GalleryConfig
import com.vluver.beta.R
import com.vluver.beta.adapter.MultiAddImages
import com.vluver.beta.model.AddMoreImages
import com.vluver.beta.servicebackground.UploadImagesPost
import com.vluver.beta.utils.GlideLoadImages
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.util.*


class PostActivity : AppCompatActivity() {
    internal lateinit var progressDialog: ProgressDialog
    internal lateinit var new_post_pictures: ImageView
    internal lateinit var regresar: ImageView
    internal lateinit var enviar: ImageView
    internal lateinit var seleccionar_imagen: ImageView
    private var user_image: ImageView? = null
    internal lateinit var descripcion_post: EditText
    internal var imageIsSet = false
    internal var bitmappost: Bitmap? = null
    private var userId: String? = null
    internal var mQueue: RequestQueue? = null


    //firebase
    private var mAuth: FirebaseAuth? = null
    var currentUser: FirebaseUser? = null
    private val channelId = "vluver_id"


    internal var PICK_IMAGE_MULTIPLE = 1
    internal var imagesEncodedList: List<String>? = null
    lateinit var mRVFish: RecyclerView
    lateinit var mAdapter: MultiAddImages
    internal lateinit var dataa: MutableList<AddMoreImages>
    internal lateinit var mGridLayoutManager: GridLayoutManager

    private var mPhotos: ArrayList<String>? = null
    internal lateinit var encodedImageList: ArrayList<String>


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_post)
        mAuth = FirebaseAuth.getInstance()
        mQueue = Volley.newRequestQueue(this@PostActivity)
        progressDialog = ProgressDialog(this)
        new_post_pictures = findViewById<View>(R.id.new_post_picture) as ImageView
        regresar = findViewById<View>(R.id.back) as ImageView
        enviar = findViewById<View>(R.id.send) as ImageView
        descripcion_post = findViewById<View>(R.id.descripcion) as EditText
        user_image = findViewById<View>(R.id.img_avatar) as ImageView
        seleccionar_imagen = findViewById<View>(R.id.new_post_submit) as ImageView
        regresar.setOnClickListener { finish() }
        mGridLayoutManager = GridLayoutManager(this@PostActivity, 3)
        dataa = ArrayList()
        encodedImageList = ArrayList()
        mRVFish = findViewById<View>(R.id.multimages) as RecyclerView
        mAdapter = MultiAddImages(this@PostActivity, dataa)
        mRVFish.adapter = mAdapter
        mRVFish.layoutManager = mGridLayoutManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 2)
            }
        }
        seleccionar_imagen.setOnClickListener {
            val config = GalleryConfig.Build()
                    .limitPickPhoto(99)
                    .singlePhoto(false)
                    .hintOfPick("max 99")
                    .filterMimeTypes(arrayOf("image/*"))
                    .build()
            GalleryActivity.openActivity(this@PostActivity, 2, config)
        }

        enviar.setOnClickListener {
            if (imageIsSet) {
                val mServiceIntent = Intent(this@PostActivity, UploadImagesPost::class.java)
                mServiceIntent.putExtra("userId", userId)
                mServiceIntent.putStringArrayListExtra("mPhotos", encodedImageList)
                mServiceIntent.putExtra("descripcion", descripcion_post.text.toString().trim { it <= ' ' })
                startService(mServiceIntent)
                finish()
            } else {
                if (descripcion_post.text.toString().isEmpty()) {
                    descripcion_post.error = "Agrega algo a tus publicacion..."
                } else {
                    insertPostToVluver(userId, "null", descripcion_post.text.toString())
                }
            }
        }
        getAvatar()

    }

    private fun getAvatar() {
        val user = FirebaseAuth.getInstance().currentUser
        if (user != null) {
            for (profile in user.providerData) {
                if (profile.providerId == "facebook.com") {
                    val image = "https://graph.facebook.com/" + user.providerData[1].uid + "/picture?type=small"
                    GlideLoadImages.loadAvatar(this@PostActivity, image, user_image)
                } else if (profile.providerId == "google.com") {
                    Toast.makeText(this@PostActivity, "avatar Google", Toast.LENGTH_SHORT).show()
                }
            }

        }
    }

    private fun sendPost(description: String) {
        progressDialog.setMessage("\tPublicando...")
        progressDialog.setCancelable(false)
        progressDialog.show()
        //insertPostToVluver(userId,"",description);
    }

    private fun insertPostToVluver(creator: String?, codeuniqueimages: String, description: String) {
        val url = "https://vluver.com/mobile/insert/insertPost.php"
        val strReq = object : StringRequest(Request.Method.POST,
                url, Response.Listener { response ->
            try {
                val jsonObject = JSONObject(response)
                val error = jsonObject.getBoolean("error")
                if (!error) {
                    // progressDialog.dismiss();
                    Toast.makeText(this@PostActivity, "Publicacion subida correctamente", Toast.LENGTH_LONG).show()
                    //show_notification("Vluver", "Publicacion lista", 2 );
                    finish()
                } else {
                    //progressDialog.dismiss();
                    val errorMsg = jsonObject.getString("error_msg")
                    Toast.makeText(this@PostActivity, "" + errorMsg, Toast.LENGTH_SHORT).show()
                }

            } catch (e: JSONException) {
                // progressDialog.dismiss();
                Toast.makeText(this@PostActivity, "Algo anda mal$e", Toast.LENGTH_SHORT).show()
                e.printStackTrace()
            }
        }, Response.ErrorListener { error ->
            //progressDialog.dismiss();
            Toast.makeText(applicationContext, error.message, Toast.LENGTH_LONG).show()
        }) {

            override fun getParams(): Map<String, String> {
                val params = HashMap<String, String>()
                params["creator"] = creator!!
                params["namecode"] = codeuniqueimages
                params["description"] = description
                return params
            }


        }
        strReq.retryPolicy = DefaultRetryPolicy(DefaultRetryPolicy.DEFAULT_TIMEOUT_MS, 5, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        mQueue!!.add(strReq)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        try {
            if (data != null) {
                mPhotos = data.getSerializableExtra(GalleryActivity.PHOTOS) as ArrayList<String>
                imageIsSet = true
                for (photo in mPhotos!!) {
                    val uri = Uri.fromFile(File(photo))
                    val addMoreImages = AddMoreImages()
                    addMoreImages.mArrayUri = uri
                    dataa.add(addMoreImages)
                    encodedImageList.add(photo)
                }
                mAdapter.notifyDataSetChanged()
                mRVFish.adapter = mAdapter
            } else {
                imageIsSet = false
                Toast.makeText(this, "Ninguna foto seleccionada!", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            imageIsSet = false
            Toast.makeText(this, "Error de seleccion: $e", Toast.LENGTH_SHORT).show()
        }


    }


    private fun hasGalleryPermission(): Boolean {
        return ActivityCompat.checkSelfPermission(this,
                android.Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
    }

    private fun askForGalleryPermission() {
        ActivityCompat.requestPermissions(this,
                arrayOf(android.Manifest.permission.READ_EXTERNAL_STORAGE),
                REQUEST_CODE_READ_PERMISSION)
    }


    override fun onStart() {
        super.onStart()
        currentUser = mAuth!!.currentUser
        userId = currentUser!!.uid
    }

    override fun onDestroy() {
        bitmappost = null
        mQueue!!.stop()
        mQueue = null
        super.onDestroy()
    }

    companion object {
        private val REQUEST_CODE_READ_PERMISSION = 22


        fun hideKeyboard(activity: Activity) {
            val imm = activity.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
            //Find the currently focused view, so we can grab the correct window token from it.
            var view = activity.currentFocus
            //If no view currently has focus, create a new one, just so we can grab a window token from it
            if (view == null) {
                view = View(activity)
            }
            assert(imm != null)
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }
}
